[SPARK-19234][MLLib] AFTSurvivalRegression should fail fast when any labels are zero

admackin opened the pull request:
## What changes were proposed in this pull request?

If any labels of 0.0 (which are invalid) are supplied, AFTSurvivalRegression gives an error straight away rather than hard-to-interpret warnings and zero-valued coefficients in the output.

## How was this patch tested?

Verified against current test suite. (One test needed to be updated as it was providing values of zero for labels so was failing after this patch)

Please review http://spark.apache.org/contributing.html before opening a pull request.

admackin committed:
[SPARK-19234][MLLib] make sure label is positive for AFT regression

admackin committed:
[SPARK-19234][MLLib] fix test suite to ensure no zero-labels get passed in test cases as they now throw errors

AmplabJenkins commented:
Can one of the admins verify this patch?

srowen commented on mllib/src/test/scala/org/apache/spark/ml/regression/AFTSurvivalRegressionSuite.scala:
```diff
@@ -18,20 +18,26 @@
 package org.apache.spark.ml.regression
 
 import scala.util.Random
-
```
(Leave the blank)

srowen replied:
This is still causing a style check failure @admackin

srowen commented on mllib/src/test/scala/org/apache/spark/ml/regression/AFTSurvivalRegressionSuite.scala:
```diff
@@ -415,4 +421,40 @@ object AFTSurvivalRegressionSuite {
     "tol" -> 0.01
   )
+
+  private[AFTSurvivalRegressionSuite] def checkNumericTypes[M <: Model[M], T <: Estimator[M]](
```
This is being copied in from MLUtils? why is it necessary?

admackin commented:
Yes, the version in MLUtils had labels of zero in the test cases, so was causing test cases to fail after my patch. It didn't look like there was a way to fix this, so I thought it better to make a patch that didn't affect potentially dozens of other packages. Any other thoughts on how to achieve this? I could add a 'minLabel' param to the MLUtils methods but that seems overly specific for this one package.

srowen commented:
I imagine there's a better way to do this without copying code. Do you mean the common test code assumes 0 labels are permitted? then maybe it should just not do that, because it's just using any old value to test.

If it doesn't actually assume 0 labels are permitted, then its logic should still work. It's just that this test would need some additional logic to verify that 0 labels cause an exception.

admackin commented:
Yes that’s right, in MLUtils it supplies zero-labels as test cases (ie assumes they’re allowed, which for other regression algorithms would be true). More than happy to patch that instead though if you think that’s OK – it’ll just affect presumably a lot of other test cases.

Verifying the 0-labels cause exceptions isn’t yet covered, but probably should be (separately to MLUtils). (I’ll just need to find out how to assert thrown errors in the Spark testing)

On 22 Jan 2017, 01:07 +1100, Sean Owen <notifications@github.com>, wrote:
> I imagine there's a better way to do this without copying code. Do you mean the common test code assumes 0 labels are permitted? then maybe it should just not do that, because it's just using any old value to test.
> If it doesn't actually assume 0 labels are permitted, then its logic should still work. It's just that this test would need some additional logic to verify that 0 labels cause an exception.
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.((

srowen commented:
It looks like there is no particular reason that `("0", Vectors.dense(0, 2, 3), 0.0)` really needs to set a 0 value as the label. You could make it 1 and see that it still passes all tests.

SparkQA commented:
**[Test build #3544 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3544/testReport)** for PR 16652 at commit [`b07c281`](https://github.com/apache/spark/commit/b07c281c378d68d86b81498ca247c7346719973e).

SparkQA commented:
**[Test build #3544 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3544/testReport)** for PR 16652 at commit [`b07c281`](https://github.com/apache/spark/commit/b07c281c378d68d86b81498ca247c7346719973e).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

Commit:
[SPARK-19234][MLLib] added test case to ensure fast failure on zero labels in AFT

admackin commented:
I've addressed all the problems I think – code style now fixed, MLTestingUtils patched (and verified all MLLib test cases still pass), and added a test case for zero-valued labels

srowen commented:
This is looking OK to me, but it needs a (squash, optionally, and) rebase now before it can test again.

SparkQA commented:
**[Test build #3551 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3551/testReport)** for PR 16652 at commit [`cfdd286`](https://github.com/apache/spark/commit/cfdd28649b1d83bfcb63cfc7b5389470125d2cad).

SparkQA commented:
**[Test build #3551 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3551/testReport)** for PR 16652 at commit [`cfdd286`](https://github.com/apache/spark/commit/cfdd28649b1d83bfcb63cfc7b5389470125d2cad).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

imatiach-msft commented on mllib/src/test/scala/org/apache/spark/ml/regression/AFTSurvivalRegressionSuite.scala:
```diff
@@ -400,6 +400,17 @@ class AFTSurvivalRegressionSuite
+         (0.000, 0.0, Vectors.dense(0.346, 2.158)), // ← generates error; zero labels invalid
+         (4.199, 0.0, Vectors.dense(0.795, -0.226)))).toDF("label", "censor", "features")
+    val aft = new AFTSurvivalRegression()
+    intercept[SparkException] {
```
it's recommended to verify the error message using withClue, eg:
withClue("label of AFTPoint must be positive") {
     intercept[SparkException] { 
        aft.fit(dataset)
     }
}

imatiach-msft commented:
looks good to me too.  I just added a small suggestion.  Thanks!

Commit:
[SPARK-19234][MLLib] added a clue about test failure per code-review suggestion

Commit:
[SPARK-19234][MLLib] fixing indentation to match code guidelines

Commit:
[SPARK-19234][MLLib] fixing imports to match scala style guidelines

SparkQA commented:
**[Test build #3552 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3552/testReport)** for PR 16652 at commit [`c855976`](https://github.com/apache/spark/commit/c855976dd2806cd7e8fd13f46d120f229f7ff6cd).

SparkQA commented:
**[Test build #3552 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3552/testReport)** for PR 16652 at commit [`c855976`](https://github.com/apache/spark/commit/c855976dd2806cd7e8fd13f46d120f229f7ff6cd).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
**[Test build #3554 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3554/testReport)** for PR 16652 at commit [`c855976`](https://github.com/apache/spark/commit/c855976dd2806cd7e8fd13f46d120f229f7ff6cd).

SparkQA commented:
**[Test build #3554 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3554/testReport)** for PR 16652 at commit [`c855976`](https://github.com/apache/spark/commit/c855976dd2806cd7e8fd13f46d120f229f7ff6cd).
 * This patch **fails PySpark unit tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

srowen commented:
The core change is looking OK @admackin but seems like it fails Python tests? if you have a moment to look at that, could be all that's needed to get this over the line.

HyukjinKwon commented:
gentle ping @admackin
